gaspy/atoms_operators.py

```diff
@@ -251,6 +252,172 @@ def find_adsorption_sites(atoms):
+            cn_dict[elem].append(cn)
+    return cn_dict
+
+
+def find_surface_atoms_indices(bulk_cn_dict, atoms):
+    '''
+    A helper function referencing codes from pymatgen to
+    get a list of surface atoms indices of a slab's
+    top surface. Due to how our workflow is setup, the
+    pymatgen method cannot be directly applied.
+
+    Taken from pymatgen.core.surface Class Slab,
+    `get_surface_sites`.
+    https://pymatgen.org/pymatgen.core.surface.html
+
+    Arg:
+        bulk_cn_dict    A dictionary of coordination numbers
+                        for each distinct site in the respective bulk structure
+        atoms           The slab where you are trying to find surface sites in
+                        `ase.Atoms` format
+    Output:
+        indices_list    A list that contains the indices of
+                        the surface atoms
+    '''
+    struct = AseAtomsAdaptor.get_structure(atoms)
+    v = VoronoiNN()
+    # Identify index of the surface atoms
+    indices_list = []
+    weights = [s.species.weight for s in struct]
+    center_of_mass = np.average(struct.frac_coords,
+                                weights=weights, axis=0)
+
+    for idx, site in enumerate(struct):
+        if site.frac_coords[2] > center_of_mass[2]:
+            try:
+                cn = v.get_cn(struct, idx, use_weights=True)
+                cn = float('%.5f' %(round(cn, 5)))
+                # surface atoms are undercoordinated
+                if cn < min(bulk_cn_dict[site.species_string]):
+                    indices_list.append(idx)
+            except RuntimeError:
+                # or if pathological error is returned,
+                # indicating a surface site
+                indices_list.append(idx)
+    return indices_list
+
+
+def _plane_normal(coords):
+    """
+    Return the surface normal vector to a plane of best fit
+    by performing planar regression.
+    See https://gist.github.com/amroamroamro/1db8d69b4b65e8bc66a6
+    for the method.
+
+    Arg:
+        coords   A `numpy.ndarray` (n,3),
+                 coordinates of atoms on the slab surface.
+
+    Output:
+        vector  numpy.ndarray. Adsorption vector for an adsorption site.
+    """
+    A = np.c_[coords[:, 0], coords[:, 1], np.ones(coords.shape[0])]
+    vector, _, _, _ = scipy.linalg.lstsq(A, coords[:, 2])
+    vector[2] = -1.0
+    vector /= -np.linalg.norm(vector)
+    return vector
+
+def _ang_between_vectors(v1, v2):
+  
+    """
+    Returns the angle in degree
+    between 3D vectors 'v1' and 'v2'
+
+    Arg:
+        v1    3D vector in np.array(x1,y1,z1) form,
+              the origin is (0,0,0).
+        v1    3D vector in np.array(x2,y2,z2) form,
+              the origin is (0,0,0).
+
+    Output:
+        angle  angle in degrees.
+    """
+    cosang = np.dot(v1, v2)
+    sinang = np.linalg.norm(np.cross(v1, v2))
+    # np.arctan2(sinang, cosang) is angle in radian
+    radian = np.arctan2(sinang, cosang)
+    angle = radian * 57.2958
```

Comment:
There's probably some function in scipy to do this more accurately. Not a big deal though and should probably keep it since we have a lot of the catalog built.